Add libvirt exporter service

chanchiwai-ray opened the pull request:
As part of the COS integration, this PR adds the [libvirt exporter](https://github.com/Tinkoff/libvirt-exporter/tree/master) service to the snap, and also add snap configuration option to enable / disable exporter services. By default, the exporter services are disabled until the `monitoring.enable` is set to `true`.

Related to: https://review.opendev.org/c/openstack/charm-openstack-hypervisor/+/893007

chanchiwai-ray committed:
Add libvirt-exporter service for exporting libvirt related metrics.

chanchiwai-ray committed:
Add snap config option to enable or disable monitoring service.

wolsen commented:
I'm struggling to understand why the exporter would be disabled by default. Libvirt services are started by default and I see no immediate reason to disable the service by default. It is true it takes up more memory, which may be a challenge for memory constrained environments but I think those environments can choose to turn it off.

chanchiwai-ray commented:
The exporters services are only part of the COS integration, and will work together with the `monitoring` plugin in sunbeam (WIP). Ideally, the exporter services should not even exist if we do not enable the `monitoring` plugin via `sunbeam enable monitoring` (WIP). That's why we choose to only bundle the service in the snap without starting it. Then, when we actually have the `monitoring` plugin, we can start the exporter services. This is the time when the exporters are useful.

Currently, there is only one exporter, but we are going to have ovs exporter in the later PR.

javacruft commented:
Irrespective of how monitoring gets integrated into Sunbeam I think its fine to have the monitoring components of the snap disabled by default - the scope of metrics exported will expand past this first component and each one exposes a new network service (and potential attack surface to the machine).

I think as a general policy an optional feature of the snap (and I think monitoring is one of them) should install disabled by default.

wolsen commented:
> I think as a general policy an optional feature of the snap (and I think monitoring is one of them) should install disabled by default.

If it should be disabled by default, that should be expressed in the snap service definitions by specifying [install-mode: disable](https://snapcraft.io/docs/services-and-daemons) for the service definition.

wolsen commented:
> > I think as a general policy an optional feature of the snap (and I think monitoring is one of them) should install disabled by default.
> 
> If it should be disabled by default, that should be expressed in the snap service definitions by specifying [install-mode: disable](https://snapcraft.io/docs/services-and-daemons) for the service definition.

And it is in this pull request. Apologies, I need more coffee.

javacruft commented on snap/snapcraft.yaml:
```diff
@@ -575,6 +586,26 @@ parts:
+      - libvirt
+    build-packages:
+      - pkg-config
+      - libvirt-dev
```
As we build libvirt from source in a previous part can we use the built artefacts from that step rather than using libvirt-dev - I'm concerned we'll end up with a conflict of some sort between the libvirt package components and the built versions.

chanchiwai-ray replied:
Thanks for the comment. Indeed, that might be a concern. 

I've looked into the source of `libvirt`, and I found that the source should already contain all the necessary files and shared library to build the libvirt exporter. I updated my patch to build the exporter with appropriate `CGO_*` flags.

hemanthnakkina committed:
Add new service ceilometer-agent-compute

Add new service to run ceilometer agent polling
for compute node.
Add ceilometer configuration and polling template files.
Update nova configuration template to include oslo
notification messaging driver.
Add config to enable/disable the service and telemetry
publish secret key.

chanchiwai-ray committed:
Optimized the build of libvirt exporter.

Commit:
Merge pull request #1 from hemanthnakkina/telemetry

Add new service ceilometer-agent-compute
